Assets/MRTK/Providers/LeapMotion/LeapMotionArticulatedHand.cs

```diff
@@ -0,0 +1,299 @@
+                case TrackedHandJoint.PinkyTip: return AttachmentPointFlags.PinkyTip;
+
+                // Metacarpals are not included in AttachmentPointFlags
+                default: return AttachmentPointFlags.Wrist;
+            }
+        }
+
+        public void UpdateState()
+        {
+            // Populate the jointPoses 
+            for (int i = 0; i < jointCount; i++)
+            {
+                TrackedHandJoint handJoint = (TrackedHandJoint)i;
+                TryGetJoint(handJoint, out MixedRealityPose pose);
+
+                if (!jointPoses.ContainsKey(handJoint))
+                {
+                    jointPoses.Add(handJoint, pose);
+                }
+                else
+                {
+                    jointPoses[handJoint] = pose;
+                }
+            }
+
+            // Update hand joints and raise event via handDefinition
+            handDefinition?.UpdateHandJoints(jointPoses);
+
+            UpdateInteractions();
+
+            UpdateVelocity();
+        }
+
+        protected void UpdateInteractions()
+        {
+            MixedRealityPose pointerPose = jointPoses[TrackedHandJoint.Palm];
+            MixedRealityPose gripPose = jointPoses[TrackedHandJoint.Palm];
+            MixedRealityPose indexPose = jointPoses[TrackedHandJoint.IndexTip];
+
+            // Only update the hand ray if the hand is in pointing pose
+            if (handDefinition.IsInPointingPose)
+            {
+                HandRay.Update(pointerPose.Position, GetPalmNormal(), CameraCache.Main.transform, ControllerHandedness);
+                Ray ray = HandRay.Ray;
+
+                pointerPose.Position = ray.origin;
+                pointerPose.Rotation = Quaternion.LookRotation(ray.direction);
+            }
+
+            for (int i = 0; i < Interactions?.Length; i++)
+            {
+                switch (Interactions[i].InputType)
+                {
+                    case DeviceInputType.SpatialPointer:
+                        Interactions[i].PoseData = pointerPose;
+                        if (Interactions[i].Changed)
+                        {
+                            CoreServices.InputSystem?.RaisePoseInputChanged(InputSource, ControllerHandedness, Interactions[i].MixedRealityInputAction, pointerPose);
+                        }
+                        break;
+                    case DeviceInputType.SpatialGrip:
+                        Interactions[i].PoseData = gripPose;
+                        if (Interactions[i].Changed)
+                        {
+                            CoreServices.InputSystem?.RaisePoseInputChanged(InputSource, ControllerHandedness, Interactions[i].MixedRealityInputAction, gripPose);
+                        }
+                        break;
+                    case DeviceInputType.Select:
+                    case DeviceInputType.TriggerPress:
+                        Interactions[i].BoolData = IsPinching;
```

Comment:
`Leap.Hand` defines things like `pinchStrength` and `grabStrength`. Do we want to try to use that data here for select and grab, instead of defining our own?

Reply:
Using the `Leap.Hand` `pinchStrength` was considered, but `isPinching` returns true for each finger and not just the index finger and the thumb.  Below is what is registered as a selection using Leap's `isPinching`.  

![LeapMotionPinchDetector](https://user-images.githubusercontent.com/53493796/79283941-32159900-7e6e-11ea-9ada-a7eef7ca150c.gif)

The Air Tap/Pinch gesture in MRTK only takes the index finger and the thumb into consideration. 

Also the `Leap.Hand` `grabStrength` is only 1 if all fingers are together, shown in the gif below:

![LeapMotionGrabDetector](https://user-images.githubusercontent.com/53493796/79286125-7441d900-7e74-11ea-8de3-9763226a8699.gif)

In MRTK, the grab gesture does not require all fingers together.  A grab can be recognized with just the thumb and the index finger.

Reply:
Interesting to know to specifics of what those APIs provide! Thanks!

>Also the Leap.Hand grabStrength is only 1 if all fingers are together, shown in the gif below

Do you think the data is "good enough" to use >0 as a grab? Or some threshold?

Reply:
Also, did you consider `pinchDistance`?

>float PinchDistance
>
>The distance between the thumb and index finger of a pinch hand pose.
>
>The distance is computed by looking at the shortest distance between the last 2 phalanges of the thumb and those of the index finger. This pinch measurement only takes thumb and index finger into account.

Reply:
The leap `pinchDistance` can be used, as IsPinching uses the same distance measurements. 

If we use greater than 0 as a grab, we might run into grab gestures getting triggered with a hand that is in a relaxed position (i.e. fingers not completely apart).  The `grabStrength` is usually always greater than 0.